docs/source/byo/interaction.rst

```diff
@@ -0,0 +1,145 @@
+Bring Your Own Interaction
+==========================
+This is a tutorial about how to implement your own interactions
+(also known as scoring functions) as subclasses of
+:class:`pykeen.nn.modules.Interaction` for use in PyKEEN.
+Imagine you've taken a time machine back to 2013 and you have just
+invented TransE [bordes2013]_, defined as:
+
+.. math::
+
+    f(h, r, t) = -\| \mathbf{e}_h + \mathbf{r}_r - \mathbf{e}_t \|_2^2
+
+where $\mathbf{e}_i$ is the $d$-dimensional embedding for entity $i$,
+$\mathbf{r}_j$ is the $d$-dimensional embedding for relation $j$, and
+$\|...\|_2$ is the $L_2$ norm.
+
+To implement TransE in PyKEEN, you need to subclass the
+:class:`pykeen.nn.modules.Interaction`. This class it itself
+a subclass of :class:`torch.nn.Module`, which means that you need to provide
+an implementation of :meth:`torch.nn.Module.forward`. However, the arguments
+are predefined as ``h``, ``r``, and ``t``, which correspond to the embeddings
+of the head, relation, and tail, respectively.
+
+.. code-block:: python
+
+    from pykeen.nn.modules import Interaction
+
+    class TransEInteraction(Interaction):
+        def forward(self, h, r, t):
+            return -(h + r - t).norm(p=2, dim=-1)
+
+Note the ``dim=-1`` because this operation is actually defined over
+an entire batch of head, relation, and tail embeddings.
+
+As another example, let's try DistMult [yang2014]_, defined as:
+
+.. math::
+
+    f(h, r, t) = <\mathbf{e}_h, \mathbf{r}_r, \mathbf{e}_t>
+
+where $\mathbf{e}_i$ is the $d$-dimensional embedding for entity $i$,
+$\mathbf{r}_j$ is the $d$-dimensional embedding for relation $j$, and
+$<x,y,z>$ is the tensor product for $x,y,z \in \mathcal{R}^d$.
```

Comment:
I am not sure whether this is actually the [tensor product](https://en.wikipedia.org/wiki/Tensor_product). iirc, in the original paper they referred to the model also as Bilinear-Diag. Alternatively, it is the (squared L2) norm of the elementwise product.

Reply:
ooof. I just want this example to be as simple as possible. I guess there are some tricks in the definition that make it much more easy to calculate. Maybe I should remove this example completely?

Reply:
See update in d5642705

Reply:
They also discuss its similarity to `TransE` (which they also name `DistAdd`): Both are the norm of an elementwise composition of the three tensors: `TransE` uses addition, whereas `DistMult` uses multiplication.

Reply:
Maybe we can write the story like this: if distmult did not exist yet, wouldn't it be an interesting variation, where we replace `+` by `*` ?

Reply:
Yes I like the thought that this could be sort of an obvious thing. Would you mind changing the sentence yourself please?

Reply:
wait nevermind. I got it.

Reply:
Done in 3abf9f6f